Mathlib/LinearAlgebra/Matrix/GeneralLinearGroup/FinTwo.lean

```diff
@@ -0,0 +1,188 @@
+  · exact fun hm ↦ ⟨_, _, (add_sub_cancel ..).symm, sub_ne_zero.mpr fun h ↦ hm.1 ⟨_, h.symm⟩,
+      hm.sub_eigenvalue_sq_eq_zero⟩
+  · rintro ⟨a, n, hm, hn0, hnsq⟩
+    constructor
+    · refine fun ⟨b, hb⟩ ↦ hn0 ?_
+      rw [← sub_eq_iff_eq_add'] at hm
+      simpa only [← hm, ← hb, ← map_sub, ← map_pow, ← map_zero (scalar (Fin 2)), scalar_inj,
+        sq_eq_zero_iff] using hnsq
+    · suffices scalar (Fin 2) (m.disc / 4) = 0 by
+        rw [← map_zero (scalar (Fin 2)), scalar_inj, div_eq_zero_iff] at this
+        have : (4 : K) ≠ 0 := by simpa [show (4 : K) = 2 ^ 2 by norm_num] using NeZero.ne _
+        tauto
+      rw [← sub_scalar_sq_eq_disc, hm, trace_add, scalar_apply, trace_diagonal]
+      simp [mul_div_cancel_left₀ _ (NeZero.ne (2 : K)),
+        (Matrix.isNilpotent_trace_of_isNilpotent ⟨2, hnsq⟩).eq_zero , hnsq]
+
+end Field
+
+section LinearOrderedRing
+
+variable {R : Type*} [CommRing R] [Preorder R]
+  (m : Matrix (Fin 2) (Fin 2) R) (g : GL (Fin 2) R)
+
+/-- A `2 × 2` matrix is *hyperbolic* if its discriminant is strictly positive. -/
+def IsHyperbolic : Prop := 0 < m.disc
+
+/-- A `2 × 2` matrix is *elliptic* if its  discriminant is strictly negative. -/
+def IsElliptic : Prop := m.disc < 0
+
+variable {m}
+
+lemma isHyperbolic_conj_iff : (g * m * g⁻¹).IsHyperbolic ↔ m.IsHyperbolic := by
+  simp only [IsHyperbolic, disc_conj]
+
+lemma isHyperbolic_conj'_iff : (g⁻¹ * m * g).IsHyperbolic ↔ m.IsHyperbolic := by
+  simpa using isHyperbolic_conj_iff g⁻¹
+
+lemma isElliptic_conj_iff : (g * m * g⁻¹).IsElliptic ↔ m.IsElliptic := by
+  simp only [IsElliptic, disc_conj]
+
+lemma isElliptic_conj'_iff : (g⁻¹ * m * g).IsElliptic ↔ m.IsElliptic := by
+  simpa using isElliptic_conj_iff g⁻¹
+
+end LinearOrderedRing
+
+namespace GeneralLinearGroup
+
+variable {R K : Type*} [CommRing R] [Field K]
+
+/-- Synonym of `Matrix.IsParabolic`, for dot-notation. -/
+@[reducible] def IsParabolic (g : GL (Fin 2) R) : Prop := g.val.IsParabolic
+
+/-- Synonym of `Matrix.IsElliptic`, for dot-notation. -/
+@[reducible] def IsElliptic [Preorder R] (g : GL (Fin 2) R) : Prop := g.val.IsElliptic
+
+/-- Synonym of `Matrix.IsHyperbolic`, for dot-notation. -/
+@[reducible] def IsHyperbolic [Preorder R] (g : GL (Fin 2) R) : Prop := g.val.IsHyperbolic
+
+/-- Polynomial whose roots are the fixed points of `g` considered as a Möbius transformation. -/
+noncomputable def fixpointPolynomial (g : GL (Fin 2) R) : R[X] :=
+  C (g 1 0) * X ^ 2 + C (g 1 1 - g 0 0) * X - C (g 0 1)
+
+/-- The fixed-point polynomial is identically zero iff `g` is scalar. -/
+lemma fixpointPolynomial_eq_zero_iff {g : GL (Fin 2) R} :
+    g.fixpointPolynomial = 0 ↔ g.val ∈ Set.range (scalar _) := by
+  rw [fixpointPolynomial]
+  constructor
+  · refine fun hP ↦ ⟨g 0 0, ?_⟩
+    have hb : g 0 1 = 0 := by simpa using congr_arg (coeff · 0) hP
+    have hc : g 1 0 = 0 := by simpa using congr_arg (coeff · 2) hP
+    have hd : g 1 1 = g 0 0 := by simpa [sub_eq_zero] using congr_arg (coeff · 1) hP
```

Comment:
I'd love if we could find a convenient way to write `a` instead of `g 0 0` etc. in situations like this. We could attempt a pattern like this to help in proofs at least:
```suggestion
  · obtain ⟨g, -, -, -⟩ := g
    obtain ⟨a, b, c, d, rfl⟩ : ∃ a b c d, !![a, b; c, d] = g :=
      ⟨g 0 0, g 1 0, g 0 1, g 1 1, sorry /- trivial missing API -/⟩
    refine fun hP ↦ ⟨a, ?_⟩
    have hb : b = 0 := by simpa using congr_arg (coeff · 0) hP
    have hc : c = 0 := by simpa using congr_arg (coeff · 2) hP
    have hd : d = a := by simpa [sub_eq_zero] using congr_arg (coeff · 1) hP
```
What do you think? It's clearly a crutch rather than a real solution so feel free to dismiss!